### PR TITLE
New fulfilment dates for Home Delivery

### PR DESCRIPTION
**Pull request by rupertbates**
## Why are you doing this?

We have a new set of criteria for Home Delivery start dates.

This table:
image (2)

shows how many days to the first delivery for each product type according to the day of the week on which the order is submitted. 

This PR then calculates start dates on the basis of that

[**Trello Card**](https://trello.com/c/73zAhO7X/2541-fix-fulfilment-delay-for-paper-hd)

### CODE REVIEW

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/deliveryDays.js [hunk 1]
```diff
@@ -61,9 +60,4 @@ const getVoucherDays = (today: number, product: ProductOptions): Date[] => {
   ).splice(weeksToAdd);
 };
 
-// TODO: this needs correcting before home delivery turned back on
-const getHomeDeliveryDays = (today: number, product: ProductOptions): Date[] =>
-  getDeliveryDays(today, getDeliveryDayForProduct(product), numberOfWeeksWeDeliverTo);
-
-
-export { getVoucherDays, getHomeDeliveryDays };
+export { getVoucherDays };
```
Comment (jfsoul) [inline support-frontend/assets/pages/paper-subscription-checkout/helpers/deliveryDays.js]: shall we rename this file to voucherDays?

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/__tests__/deliveryDays.js [hunk 2]
```diff
@@ -26,15 +29,34 @@ describe('deliveryDays', () => {
...
-      expect(formatMachineDate(days[0])).toEqual('2019-03-04');
+    it('delivers first Sixday paper 6 days later if ordered on a Thursday', () => {
+      const days = getHomeDeliveryDays(thursday, 'Sixday');
+      expect(formatMachineDate(days[0])).toEqual('2019-03-06');
+    });
+    it('delivers first Weekend paper 9 days later if ordered on a Thursday', () => {
+      const days = getHomeDeliveryDays(thursday, 'Weekend');
+      expect(formatMachineDate(days[0])).toEqual('2019-03-09');
+    });
+    it('delivers first Sunday paper 10 days later if ordered on a Thursday', () => {
```
Comment (jfsoul) [inline support-frontend/assets/pages/paper-subscription-checkout/helpers/__tests__/deliveryDays.js]: I would've put a few more cases in so we're covering more than thursday acquisitions?

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+): boolean => daysTillNextDelivery(currentWeekDay, deliveryDay) >= delayDays;
+
+const getHomeDeliveryDays = (today: number, product: ProductOptions): Date[] => {
+  const currentWeekday = new Date(today).getDay();
+  const delayDays = getDaysToAdd(currentWeekday, product);
+  const deliveryDay: Day = ((currentWeekday + delayDays) % 7: any);
+  const canMakeNextDelivery = canDeliverOnNextDeliveryDay(currentWeekday, deliveryDay, delayDays);
+  const deliveryIsThisWeek = jsDayToFulfilmentDay(deliveryDay) > jsDayToFulfilmentDay(currentWeekday);
+  const weeksToAdd = canMakeNextDelivery && deliveryIsThisWeek ? 0 : 1;
+  return getDeliveryDays(
```
Comment (jfsoul) [inline support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js]: Is this assuming that each option is 1 week apart for HD? That's the case for Saturday/Sunday but not necessarily for Everyday/Sixday/Weekend.

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/deliveryDays.js (hunk 1 above)
Comment (jfsoul) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/deliveryDays.js]: Nevermind, looks like you did this already

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  currentWeekDay: number,
+  deliveryDay: number,
+  delayDays: number,
+): boolean => daysTillNextDelivery(currentWeekDay, deliveryDay) >= delayDays;
+
+const getHomeDeliveryDays = (today: number, product: ProductOptions): Date[] => {
+  const currentWeekday = new Date(today).getDay();
+  const delayDays = getDaysToAdd(currentWeekday, product);
+  const deliveryDay: Day = ((currentWeekday + delayDays) % 7: any);
+  const canMakeNextDelivery = canDeliverOnNextDeliveryDay(currentWeekday, deliveryDay, delayDays);
```
Comment (jfsoul) [inline support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js]: Am I missing something or could we just do `const firstOption = today.addDays(delayDays)` for the first delivery option and then `daysTillNextDelivery(firstOption)` recursively for further dates?

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js (hunk 3 above)
Comment (rupertbates) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js]: No it's working out what day of the week the delivery falls on based on the number of days we need to add and then whether that takes us into next week or not. If it does then we need to ignore the first value in the array of possible dates returned by `getDeliveryDays`

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js (hunk 4 above)
Comment (rupertbates) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js]: I wanted to use the `getDeliveryDays()` function that all the other products including voucher use for consistency's sake, which means that we need to give it the delivery day of the week and then work out whether the first date it returns us is valid or not ie. whether we can fulfil by that date in that week.

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js (hunk 3 above)
Comment (rupertbates) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/homeDeliveryDays.js]: After offline discussion I understand what you're asking here now. Yes it does make that assumption so we need to work out what we should do here.

**Pull request merged**